### PR TITLE
Integration with kodi

### PR DESCRIPTION
**Pull request by albfan**
Kodi is a media center manager, to integrate with Ryujinx kodi needs:

- List of games available
- Interact without keyboard or mouse

This PR adds needed functionality, extending or modifying a bit Ryujinx internals:

- To list game list metadata, a `--list-games` option is added
- 
This will output a json array of metadata for games:

```
[
  {
    "path": "/home/user/games/foo.nsp",
    "title_name": "Foo game",
    "icon": "/9j/4QD9RXhpZgAATU0AKgAAAAg..."
  },
  {
    "path": "/home/user/games/bar.nsp",
    "title_name": "Bar game",
    "icon": "/9j/4QD9RXhpZgAATU0AKgAAAAgACAEPAAIAAAASAAAAbgESAAMAAAABAA..."
  }
]
```

that will be consumed by kodi game addon.

I'm not expert on .NET so a review is highly appreciated.

Good to have:
- Refactor of packages so this CLI function do not relay on Ryujinx.UI packages

To interact with emulation without keyboard or mouse a trick used in retroarch is added. Detect special combos on joycons to execute actions. Actions needed are:

- Wake up message (`Plus` + `Minus`)
- Exit emulation (`Plus` + `Minus` + `L` + `R`)

Modified behaviour of ryujinx is to exit Ryujinx if you stop emulation launching it with a nsp file as argument (like just run emulation and avoid to go to game list after that)

This PR is fully functional with this Ryujinx kodi addon: xbmc/repo-plugins#3510

### CODE REVIEW

**Comment by albfan** [comment]
Added some quick and dirty things to avoid stop emulation to go into game list. Probably a --once option will be interesting to do this on demand (if for any reason launch ryujinx with a nsp file you still want to see the game list after stop emulation)

To give context, this changes are a tentative design for a kodi ryujinx addon, which is ready to test here: https://github.com/xbmc/repo-plugins/pull/3510

File: Ryujinx/Program.cs [hunk 1]
```diff
@@ -128,6 +162,28 @@ static void Main(string[] args)
                 ConfigurationState.Instance.Ui.StartFullscreen.Value = true;
             }
 
+            if (listGames)
+            {
+
```
Comment (riperiperi) [inline Ryujinx/Program.cs:167]: Unnecessary newline

File: Ryujinx/Program.cs [hunk 2]
```diff
@@ -69,6 +98,11 @@ static void Main(string[] args)
             });
 
             Version = Assembly.GetEntryAssembly().GetCustomAttribute<AssemblyInformationalVersionAttribute>().InformationalVersion;
+	    if (showVersion)
```
Comment (riperiperi) [inline Ryujinx/Program.cs:101]: The indentation here is wrong, and there should be a new line above.

File: Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs [hunk 3]
```diff
@@ -100,6 +100,7 @@ public bool Validate(int playerMin, int playerMax, ControllerType acceptedTypes,
 
             if (configuredCount < playerMin || configuredCount > playerMax || primaryIndex == PlayerIndex.Unknown)
             {
+                Logger.Info?.Print(LogClass.Hid, $"min{playerMin} count {configuredCount} max {playerMax} index unknown {primaryIndex == PlayerIndex.Unknown}");
```
Comment (riperiperi) [inline Ryujinx.HLE/HOS/Services/Hid/HidDevices/NpadDevices.cs:103]: This log doesn't seem related to this PR.

File: Ryujinx/Program.cs [hunk 4]
```diff
@@ -128,6 +162,28 @@ static void Main(string[] args)
                 ConfigurationState.Instance.Ui.StartFullscreen.Value = true;
             }
 
+            if (listGames)
+            {
+
+                ArrayList gameMetadataArray = new ArrayList();
+                VirtualFileSystem virtualFileSystem      = VirtualFileSystem.CreateInstance();
```
Comment (riperiperi) [inline Ryujinx/Program.cs:169]: Alignment doesn't match other variables.

File: Ryujinx/Program.cs [hunk 5]
```diff
@@ -128,6 +162,28 @@ static void Main(string[] args)
                 ConfigurationState.Instance.Ui.StartFullscreen.Value = true;
             }
 
+            if (listGames)
+            {
+
+                ArrayList gameMetadataArray = new ArrayList();
```
Comment (riperiperi) [inline Ryujinx/Program.cs:168]: Add a newline under this, or align it with the other variables.

File: Ryujinx/Program.cs [hunk 6]
```diff
@@ -128,6 +162,28 @@ static void Main(string[] args)
...
+                    {
+                        { "path", args.AppData.Path },
+                        { "titleName", args.AppData.TitleName },
+                        { "icon", Convert.ToBase64String(args.AppData.Icon, 0, args.AppData.Icon.Length, Base64FormattingOptions.InsertLineBreaks) }
+                    };
+                    gameMetadataArray.Add(data);
+                };
+                applicationLibrary.LoadApplications(ConfigurationState.Instance.Ui.GameDirs, ConfigurationState.Instance.System.Language);
+                Console.Write(JsonHelper.Serialize(gameMetadataArray, true));
+                return;
```
Comment (riperiperi) [inline Ryujinx/Program.cs:184]: Newline before return.

File: Ryujinx/Program.cs [hunk 7]
```diff
@@ -128,6 +162,28 @@ static void Main(string[] args)
...
             }
 
+            if (listGames)
+            {
+
+                ArrayList gameMetadataArray = new ArrayList();
+                VirtualFileSystem virtualFileSystem      = VirtualFileSystem.CreateInstance();
+                ContentManager contentManager         = new ContentManager(virtualFileSystem);
+                ApplicationLibrary applicationLibrary = new ApplicationLibrary(virtualFileSystem);
+                applicationLibrary.ApplicationAdded  += delegate(object sender, ApplicationAddedEventArgs args)
```
Comment (riperiperi) [inline Ryujinx/Program.cs:172]: Newline before and after this event assignment, so that it stands out more. You should also remove the alignment before `+=` if you do this.

File: Ryujinx/Ui/GLRenderer.cs [hunk 8]
```diff
@@ -614,6 +615,16 @@ private bool UpdateFrame()
                     };
                 }
 
+                ControllerKeys specialCombo1 = ControllerKeys.Minus|ControllerKeys.L|ControllerKeys.Plus|ControllerKeys.R;
```
Comment (riperiperi) [inline Ryujinx/Ui/GLRenderer.cs:618]: These should be const, and defined at the top of the class. Their names should also be more descriptive.

File: Ryujinx/Ui/GLRenderer.cs [hunk 9]
```diff
@@ -37,6 +37,7 @@ static GlRenderer()
         public ManualResetEvent WaitEvent { get; set; }
 
         public static event EventHandler<StatusUpdatedEventArgs> StatusUpdatedEvent;
+        public event EventHandler<ComboPressedEventArgs>         ComboPressed;
```
Comment (riperiperi) [inline Ryujinx/Ui/GLRenderer.cs:40]: The name `ComboPressed` doesn't really indicate what the combos are or what they're for.

File: Ryujinx/Ui/GLRenderer.cs [hunk 10]
```diff
@@ -614,6 +615,16 @@ private bool UpdateFrame()
                     };
                 }
 
+                ControllerKeys specialCombo1 = ControllerKeys.Minus|ControllerKeys.L|ControllerKeys.Plus|ControllerKeys.R;
+                ControllerKeys specialCombo2 = ControllerKeys.Minus|ControllerKeys.Plus;
+
+                if ((currentButton & specialCombo1) == specialCombo1) {
```
Comment (riperiperi) [inline Ryujinx/Ui/GLRenderer.cs:621]: `{` should be on new line.

File: Ryujinx/Ui/GLRenderer.cs [hunk 11]
```diff
@@ -614,6 +615,16 @@ private bool UpdateFrame()
...
+                    OnComboPressed(new ComboPressedEventArgs(1));
+                }
+                else if ((currentButton & specialCombo2) == specialCombo2) {
```
Comment (riperiperi) [inline Ryujinx/Ui/GLRenderer.cs:624]: `{` should be on new line.

File: Ryujinx/Ui/GLRenderer.cs [hunk 12]
```diff
@@ -742,5 +753,11 @@ private bool UpdateFrame()
 
             return true;
         }
+
+        protected void OnComboPressed(ComboPressedEventArgs args)
+        {
+            ComboPressed?.Invoke(null, args);
+        }
+
```
Comment (riperiperi) [inline Ryujinx/Ui/GLRenderer.cs:761]: Unnecessary new line.

File: Ryujinx/Ui/MainWindow.cs [hunk 13]
```diff
@@ -634,6 +636,10 @@ private void CreateGameWindow()
             DisplaySleep.Prevent();
 
             GlRendererWidget = new GlRenderer(_emulationContext, ConfigurationState.Instance.Logger.GraphicsDebugLevel);
+            if (runOnce)
```
Comment (riperiperi) [inline Ryujinx/Ui/MainWindow.cs:639]: Newline before.

File: Ryujinx/Ui/MainWindow.cs [hunk 14]
```diff
@@ -663,6 +669,10 @@ private void CreateGameWindow()
 
             _emulationContext.Dispose();
             _deviceExitStatus.Set();
+            if (runOnce)
```
Comment (riperiperi) [inline Ryujinx/Ui/MainWindow.cs:672]: Newline before.

File: Ryujinx/Ui/MainWindow.cs [hunk 15]
```diff
@@ -565,6 +566,7 @@ public void LoadApplication(string path)
                         case ".pfs0":
                             Logger.Info?.Print(LogClass.Application, "Loading as NSP.");
                             _emulationContext.LoadNsp(path);
+                            runOnce = true;
```
Comment (riperiperi) [inline Ryujinx/Ui/MainWindow.cs:569]: Won't this cause all runs of an NSP to exit? (even if you aren't starting a game directly)

File: Ryujinx/Ui/MainWindow.cs [hunk 16]
```diff
@@ -712,6 +722,26 @@ private void CreateGameWindow()
             });
         }
 
+        private void Combo_Pressed(object sender, ComboPressedEventArgs args)
+        {
+            if (args.Combo == 1)
```
Comment (riperiperi) [inline Ryujinx/Ui/MainWindow.cs:727]: Combos should be named using an `enum`, if you're going to add them.

**Review by riperiperi — COMMENTED** [review]
Left some comments, mostly style related. I haven't used kodi but it would be nice to have Ryu usable from interfaces like this.

I noticed that you added a `runOnce` variable that is set to true whenever an NSP is loaded. Perhaps this should only be set true when the user boots ryu directly into a game, or if they provide an argument like `--runOnce`.

**Comment by albfan** [comment]
@riperiperi thanks for review, will rework all this PR based on that.

I need the new input system from #2194 so I rebase on top of that. https://github.com/albfan/Ryujinx/tree/new-ryujinx

There're still things to fix on kodi, and some tricks we need to discuss, but will follow your guidelines and stick to code style narrowing the undefined solutions from iteration to iteration.

Will split things so we can see what can be integrated and what needs a second though.

**Comment by AcK77** [comment]
I think you should understand that the way you do things will sadly never be merged to Ryujinx as-is. You changed some code from the core of the project just to fit some Kodi requirements (like the changes in the hid service).
I suggest you to write another backend as we did for GTK or SDL2 to send all needed infos to Kodi, instead of modify things here and there which could cause issues even if you use Kodi.
I don’t want to discourage you, I’d just want to be clear with the goal of that PR!

**Comment by albfan** [comment]
I rebase this code on top of master.

I can split into small PR to decide on each feature

**Comment by marysaka** [comment]
> I rebase this code on top of master.
> 
> I can split into small PR to decide on each feature

Might be worth splitting stuffs and discuss on each individual PR if it should be integrated the way it is written 👍

**Comment by AcK77** [comment]
As @marysaka said stuffs should be splitted and discussed on each individual PRs.
Since this PR isn't maintained and nothing have been PRd, I close it.
Thanks.
